### PR TITLE
Improvements to `ip netns` code

### PR DESCRIPTION
**Pull request by mfoliveira**
This PR:
 - Fixes one issue where the networking plugin runs commands on `Error` namespaces;
 - Gathers per-namespace data in the conntrack plugin too.
 - Deduplicates code to get/filter the list of namespaces in the conntrack/ebpf/networking plugins.

Each commit message contains testing details, and a final `# ./tests/simple.sh` ends in `Everything worked!`.

Comparing the contents of `sos_commands` before/after patches, the only changes are the new per-namespace commands in the `conntrack` plugin:

Before and After:
```
$ sudo ./bin/sos report -o conntrack,ebpf,networking --batch
```
Compare:
```
$ find sosreport-<OLD>/sos_commands/ | cut -d/ -f2- | sort \
    > find-sosreport.old

$ find sosreport-<NEW>/sos_commands/ | cut -d/ -f2- | sort \
    > find-sosreport.new

$ diff find-sosreport.{old,new}
4a5,10
> sos_commands/conntrack/ip_netns_exec_ns1_conntrack_-L_-o_extended
> sos_commands/conntrack/ip_netns_exec_ns1_conntrack_-S
> sos_commands/conntrack/ip_netns_exec_ns2_conntrack_-L_-o_extended
> sos_commands/conntrack/ip_netns_exec_ns2_conntrack_-S
> sos_commands/conntrack/ip_netns_exec_ns3_conntrack_-L_-o_extended
> sos_commands/conntrack/ip_netns_exec_ns3_conntrack_-S
```
And the contents of the `ip_netns_exec` files is identical before/after:

```
$ md5sum sosreport-<OLD>/sos_commands/*/* | grep ip_netns_exec | sed 's, [^/]\+/, ,' > md5sum-sosreport.old
$ md5sum sosreport-<NEW>/sos_commands/*/* | grep ip_netns_exec | sed 's, [^/]\+/, ,' > md5sum-sosreport.new

$ diff md5sum-sosreport.{old,new}
0a1,6
> 70e61b0a6cfc43f1fbaa7884f5296495 sos_commands/conntrack/ip_netns_exec_ns1_conntrack_-L_-o_extended
> 84e5f5c039b6d1f35f836f0a2a7c4cdc sos_commands/conntrack/ip_netns_exec_ns1_conntrack_-S
> 70e61b0a6cfc43f1fbaa7884f5296495 sos_commands/conntrack/ip_netns_exec_ns2_conntrack_-L_-o_extended
> 84e5f5c039b6d1f35f836f0a2a7c4cdc sos_commands/conntrack/ip_netns_exec_ns2_conntrack_-S
> 70e61b0a6cfc43f1fbaa7884f5296495 sos_commands/conntrack/ip_netns_exec_ns3_conntrack_-L_-o_extended
> 84e5f5c039b6d1f35f836f0a2a7c4cdc sos_commands/conntrack/ip_netns_exec_ns3_conntrack_-S
```

---
Please place an 'X' inside each '[]' to confirm you adhere to our [Contributor Guidelines](https://github.com/sosreport/sos/wiki/Contribution-Guidelines)

- [x] Is the commit message split over multiple lines and hard-wrapped at 72 characters?
- [x] Is the subject and message clear and concise?
- [x] Does the subject start with **[plugin_name]** if submitting a plugin patch or a **[section_name]** if part of the core sosreport code?
- [x] Does the commit contain a **Signed-off-by: First Lastname <email@example.com>**?
- [ ] If this commit closes an existing issue, is the line `Closes: #ISSUENUMBER` included in an independent line?
- [ ] If this commit resolves an existing pull request, is the line `Resolves: #PRNUMBER` included in an independent line?

### CODE REVIEW

File: sos/utilities.py
Comment (TurboTurtle) [inline sos/utilities.py:219]: Nack. Use the logging already setup by sos. This will be available inside the `Plugin` class already.

**Review by TurboTurtle — COMMENTED** [review]
Some initial comments follow.

I like the aim of `get_network_namespaces()` however I think this should either:
1. Go into `Plugin()` rather than into `utilities.py`
2. Be called during initialization and passed to plugins via `commons` like we do for disks and fibre devices, then have a wrapper in `Plugin` to filter the cached results as needed.

I prefer option 2 personally, but I wouldn't object to option 1.

File: sos/utilities.py
Comment (mfoliveira) [reply sos/utilities.py:230]: 1. Indeed; did I get it right, here? Sending it this way for now.
```
+            pattern = (
+                '(?:%s$)' % '$|'.join(ns_pattern.split()).replace('*', '.*')
+                )
```

2. Ack; indeed, I just copied it over, trying to be conservative. :)

File: sos/utilities.py
Comment (mfoliveira) [reply sos/utilities.py:246]: Ack.

File: sos/utilities.py
Comment (mfoliveira) [reply sos/utilities.py:219]: Face palm. I missed that came from `Plugin`. Fixed it.

**Comment by mfoliveira** [comment]
@TurboTurtle 

Thanks for reviewing!

> Some initial comments follow.
> 
> I like the aim of `get_network_namespaces()` however I think this should either:
> 
>     1. Go into `Plugin()` rather than into `utilities.py`
> 
>     2. Be called during initialization and passed to plugins via `commons` like we do for disks and fibre devices, then have a wrapper in `Plugin` to filter the cached results as needed.
> 
> 
> I prefer option 2 personally, but I wouldn't object to option 1.

Option 2 indeed looks better and more consistent with what's already available.
I implemented that, along with your suggestions above. Pushing the changes.

**Comment by mfoliveira** [comment]
> This pull request **introduces 1 alert** when merging [99de64b](https://github.com/sosreport/sos/commit/99de64b29b117bfe1d92fdd672d7148a729db5e3) into [7314865](https://github.com/sosreport/sos/commit/7314865bab8569e2d6f64e0e30f4118c18bdc2f6) - [view on LGTM.com](https://lgtm.com/projects/g/sosreport/sos/rev/pr-81601b49d2c1ffdddfe380bd06135cc277bf432a)
> 
> **new alerts:**
> 
>     * 1 for Unused import

Removed `from re import match` from `networking.py`, not used anymore.

File: sos/report/plugins/__init__.py
Comment (pmoravec) [inline sos/report/plugins/__init__.py:2681]: This means, if *both* `ns_pattern` and `ns_max` is set, the later option is ignored? I am not sure how broad scenario it is to call networking plugin with both limits applied, but still I see it as a legitimate one.

Could this scenario be supported as well?

Comment (mfoliveira) [reply sos/report/plugins/__init__.py:2681]: Hi Pavel,

Thanks for reviewing.

Good point. I wondered about that as well, as to why the options are mutually exclusive, but their combination does seem  a legitimate use case, as you mentioned. (e.g., all namespaces matching a regex, but at most N, in case the system has way too many.)

I guess it can be supported, but I'd have to defer to more experienced folks/author/maintainers, as I'm fresh new to sosreport code, as you may have seen from previous suggestions. :)

For now/this PR, I focused more on keeping the current behavior (sure enough) and just changing places for current things; but would be happy to take that route / add another commit with your suggestion, if that is the consensus.

cheers,
Mauricio

File: sos/report/plugins/__init__.py
Comment (TurboTurtle) [inline sos/report/plugins/__init__.py:2665]: Take a look at the device related commands in plugin for a reference - this should be automatically using the namespaces discovered earlier rather than having the plugins pass that same value in every time.

E.G.

```python
def get_namespaces(self, ns_pattern=None, ns_max=0):
[...]
    for ns in self.commons[namespaces][network]
```

Then in the plugins you'd use `for namespace in self.get_namespaces()` for the whole list, passing only a regex if you want to filter those down.

File: sos/report/plugins/networking.py
Comment (TurboTurtle) [inline sos/report/plugins/networking.py:240]: This would all be condensed down to the following for example with the changes mentioned above

```python
for namespace in self.get_namespaces(self.get_option("namespace_pattern"),
                                     self.get_option("namespaces")):
    [... collections ...]
```

**Review by TurboTurtle — CHANGES_REQUESTED** [review]

File: sos/report/plugins/__init__.py
Comment (pmoravec) [reply sos/report/plugins/__init__.py:2681]: Ah right, now I recall there was some discussion in #1916 without a consensus. Let dont change the behaviour and just copy the current one, here.

File: sos/report/plugins/__init__.py
Comment (mfoliveira) [reply sos/report/plugins/__init__.py:2665]: I see; I tried to make the function generic in case other non-network, namespace types are implemented/used in the future.

But this approach looks much simpler for users, and it can use `get_network_namespaces()` to particularize for the network namespace type.

I'll resubmit with those changes; thanks!

File: sos/report/plugins/networking.py
Comment (mfoliveira) [reply sos/report/plugins/networking.py:240]: I was lazy and kept `if` and `out_ns` to avoid the large indentation changes, and due to the other use of `out_ns` for `ethtool_namespaces` option.
I'll update both more properly this time. :)

**Comment by mfoliveira** [comment]
@TurboTurtle 

Just pushed the changes with your suggestions, and updated testing in PR description. Thanks!

**Comment by TurboTurtle** [comment]
Code-wise this looks good to me, and a couple test runs look right.

Just 2 things:

1. One more rebase to pick up the CirrusCI changes, so this PR will then kick off a set of the automated testing
2. There are some PEP8 cleanups needed:

```
sos/report/__init__.py:20:80: E501 line too long (81 > 79 characters)
sos/report/plugins/__init__.py:2665:48: E251 unexpected spaces around keyword / parameter equals
sos/report/plugins/__init__.py:2665:50: E251 unexpected spaces around keyword / parameter equals
sos/report/plugins/__init__.py:2665:63: E251 unexpected spaces around keyword / parameter equals
sos/report/plugins/__init__.py:2665:65: E251 unexpected spaces around keyword / parameter equals
sos/report/plugins/__init__.py:2669:52: E251 unexpected spaces around keyword / parameter equals
sos/report/plugins/__init__.py:2669:54: E251 unexpected spaces around keyword / parameter equals
sos/report/plugins/__init__.py:2669:67: E251 unexpected spaces around keyword / parameter equals
sos/report/plugins/__init__.py:2669:69: E251 unexpected spaces around keyword / parameter equals
```

If Travis was working this would have been caught earlier, but this will now also be caught by Cirrus once point 1 is done.

**Comment by mfoliveira** [comment]
> Code-wise this looks good to me, and a couple test runs look right.

Cool; thanks for promptly reviewing!

> Just 2 things:
> 
>     1. One more rebase to pick up the CirrusCI changes, so this PR will then kick off a set of the automated testing
> 
>     2. There are some PEP8 cleanups needed:

Uh, I had the impression I did rebase; just re-did, and CirrusCI is running now.  And fixed the PEP8 warnings.

**Review by TurboTurtle — APPROVED** [review]

**Review by pmoravec — APPROVED** [review]

**Comment by mfoliveira** [comment]
@TurboTurtle @pmoravec 

Thanks for speedy reviews, corrections, and suggestions!

**Comment by pmoravec** [comment]
.. and thanks for the fast implementation of our feedback, though it might be seen as punctilious sometimes.

**Pull request closed without merging**
